File: telegram_monitor.py
"""
Module theo dõi Telegram để lấy referral codes
"""
import requests
import re
import time
import threading
from datetime import datetime

class TelegramMonitor:
    def __init__(self, bot_token, chat_id):
        self.bot_token = bot_token
        self.chat_id = chat_id
        self.last_update_id = 0
        self.running = False
        
    def get_updates(self):
        """Lấy tin nhắn mới từ Telegram"""
        try:
            url = f"https://api.telegram.org/bot{self.bot_token}/getUpdates"
            params = {
                'offset': self.last_update_id + 1,
                'limit': 100,
                'timeout': 30
            }
            
            response = requests.get(url, params=params, timeout=35)
            if response.status_code == 200:
                return response.json()
            else:
                print(f"❌ Lỗi Telegram API: {response.status_code}")
                return None
        except Exception as e:
            print(f"❌ Lỗi kết nối Telegram: {e}")
            return None
    
    def extract_codes(self, text):
        """Trích xuất codes 8 ký tự từ text"""
        # Pattern: 8 ký tự gồm chữ và số, không phải toàn chữ hoặc toàn số
        pattern = r'\b[A-Z0-9]{8}\b'
        matches = re.findall(pattern, text.upper())
        
        valid_codes = []
        for code in matches:
            # Loại bỏ code toàn số hoặc toàn chữ
            if not (code.isdigit() or code.isalpha()):
                valid_codes.append(code)
        
        return valid_codes
    
    def monitor(self, callback=None):
        """Theo dõi Telegram liên tục"""
        print(f"🔍 Bắt đầu theo dõi Telegram chat: {self.chat_id}")
        self.running = True
        
        while self.running:
            try:
                updates = self.get_updates()
                
                if updates and updates.get('ok'):
                    for update in updates.get('result', []):
                        self.last_update_id = update['update_id']
                        
                        if 'message' in update:
                            message = update['message']
                            
                            # Kiểm tra có phải từ chat target không
                            chat_id = str(message['chat']['id'])
                            if chat_id == self.chat_id or f"-{chat_id}" == self.chat_id:
                                
                                text = message.get('text', '')
                                timestamp = datetime.fromtimestamp(message['date']).strftime('%H:%M:%S')
                                sender = message.get('from', {}).get('first_name', 'Unknown')
                                
                                codes = self.extract_codes(text)
                                
                                if codes:
                                    print(f"\n📨 [{timestamp}] {sender}: {text[:50]}...")
                                    for code in codes:
                                        print(f"🎯 Tìm thấy code: {code}")
                                        
                                        # Callback để xử lý code
                                        if callback:
                                            callback(code)
                
                time.sleep(2)  # Delay để tránh spam API
                
            except KeyboardInterrupt:
                print("\n🔴 Dừng theo dõi Telegram...")
                self.running = False
                break
            except Exception as e:
                print(f"❌ Lỗi monitor: {e}")
                time.sleep(5)
    
    def start_monitoring(self, callback=None):
        """Bắt đầu theo dõi trong thread riêng"""
        monitor_thread = threading.Thread(target=self.monitor, args=(callback,))
        monitor_thread.daemon = True
        monitor_thread.start()
        return monitor_thread

def test_telegram():
    """Test Telegram monitor với token demo"""
    print("🔍 Test Telegram Pattern Extraction...")
    
    # Test messages
    test_messages = [
        "Hey everyone! Just a quick reminder about how we share referral codes ABC12345 and links in the community.",
        "I have code but only give ABCD1234 for you",
        "can u give me code XYZ98765",
        "Check this out: https://sodex.dev/ref/CODE123X",
        "My code is: BONUS789",  # Toàn chữ, sẽ bị loại
        "Use code: 12345678",    # Toàn số, sẽ bị loại
        "Valid codes: TEST1234, DEMO5678, WORK9ABC"
    ]
    
    monitor = TelegramMonitor("demo_token", "demo_chat")
    
    for msg in test_messages:
        codes = monitor.extract_codes(msg)
        if codes:
            print(f"📨 Message: {msg}")
            print(f"🎯 Extracted codes: {codes}")
            print("-" * 50)

if __name__ == "__main__":
    test_telegram()

File: telegram-monitor.js
const { Telegraf } = require('telegraf');
const EventEmitter = require('events');

class TelegramMonitor extends EventEmitter {
    constructor(botToken, groupId) {
        super();
        this.bot = new Telegraf(botToken);
        this.groupId = groupId;
        this.isMonitoring = false;
    }

    async start() {
        if (this.isMonitoring) return;

        console.log('🤖 Bắt đầu theo dõi Telegram...');

        // Lắng nghe tin nhắn trong group
        this.bot.on('text', (ctx) => {
            try {
                const message = ctx.message;
                const chatId = message.chat.id.toString();
                
                // Kiểm tra xem tin nhắn có từ group được chỉ định không
                if (chatId === this.groupId || message.chat.username) {
                    const text = message.text;
                    console.log(`📱 Nhận tin nhắn: ${text}`);
                    
                    // Tìm referral code 8 ký tự (có thể là chữ và số)
                    const referralCodes = this.extractReferralCodes(text);
                    
                    if (referralCodes.length > 0) {
                        referralCodes.forEach(code => {
                            console.log(`🎯 Tìm thấy referral code: ${code}`);
                            this.emit('referralCodeFound', code);
                        });
                    }
                }
            } catch (error) {
                console.error('❌ Lỗi xử lý tin nhắn Telegram:', error);
            }
        });

        // Bắt đầu bot
        await this.bot.launch();
        this.isMonitoring = true;
        
        console.log('✅ Telegram monitor đã khởi động');
    }

    extractReferralCodes(text) {
        const codes = [];
        
        // Pattern 1: Tìm code 8 ký tự được đề cập cụ thể
        const patterns = [
            /referral[:\s]*([A-Za-z0-9]{8})/gi,
            /ref[:\s]*([A-Za-z0-9]{8})/gi,
            /code[:\s]*([A-Za-z0-9]{8})/gi,
            /\b([A-Za-z0-9]{8})\b/g  // Bất kỳ chuỗi 8 ký tự nào đứng một mình
        ];

        patterns.forEach(pattern => {
            let match;
            while ((match = pattern.exec(text)) !== null) {
                const code = match[1];
                // Kiểm tra xem code có hợp lệ không (không phải toàn số hoặc toàn chữ)
                if (this.isValidReferralCode(code)) {
                    codes.push(code);
                }
            }
        });

        // Loại bỏ duplicate
        return [...new Set(codes)];
    }

    isValidReferralCode(code) {
        // Kiểm tra độ dài
        if (code.length !== 8) return false;
        
        // Kiểm tra có cả số và chữ
        const hasNumber = /\d/.test(code);
        const hasLetter = /[A-Za-z]/.test(code);
        
        // Phải có cả số và chữ (không được toàn số hoặc toàn chữ)
        return hasNumber && hasLetter;
    }

    stop() {
        if (this.bot) {
            this.bot.stop();
            this.isMonitoring = false;
            console.log('🛑 Dừng theo dõi Telegram');
        }
    }
}

module.exports = TelegramMonitor;
